pkg/mixin/search.go

```diff
@@ -0,0 +1,82 @@
+package mixin
+
+import (
+	"encoding/json"
+	"sort"
+	"strings"
+
+	"get.porter.sh/porter/pkg/printer"
+	"github.com/gobuffalo/packr/v2"
+	"github.com/pkg/errors"
+)
+
+// SearchOptions are options for searching remote mixins
+type SearchOptions struct {
+	Name string
+	printer.PrintOptions
+}
+
+// Searcher contains a packr.Box containing a searchable list of mixins
+type Searcher struct {
+	Box *packr.Box
+}
+
+// NewSearcher returns a Searcher with the provided packr.Box
+func NewSearcher(box *packr.Box) Searcher {
+	return Searcher{
+		Box: box,
+	}
+}
+
+// Validate validates the arguments provided to a search command
+func (o *SearchOptions) Validate(args []string) error {
+	err := o.validateMixinName(args)
+	if err != nil {
+		return err
+	}
+
+	return o.ParseFormat()
+}
+
+// validateMixinName validates either no mixin name is provided or only one is
+func (o *SearchOptions) validateMixinName(args []string) error {
+	switch len(args) {
+	case 0:
+		return nil
+	case 1:
+		o.Name = strings.ToLower(args[0])
+		return nil
+	default:
+		return errors.Errorf("only one positional argument may be specified, the mixin name, but multiple were received: %s", args)
+	}
+}
+
+// Search searches for mixins matching the optional provided name,
+// returning the full list if none is provided
+func (s *Searcher) Search(opts SearchOptions) (RemoteMixinList, error) {
+	data, err := s.Box.Find("index.json")
+	if err != nil {
+		return nil, errors.Wrap(err, "error loading remote mixin list")
+	}
+
+	var rml RemoteMixinList
+	err = json.Unmarshal(data, &rml)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not parse remote mixin list")
+	}
+
+	results := make(RemoteMixinList, len(rml))
+	copy(results, rml)
+
+	if opts.Name != "" {
+		results = RemoteMixinList{}
+		for _, rmi := range rml {
+			if strings.Contains(rmi.Name, opts.Name) {
+				results = append(results, rmi)
+			}
+		}
+	}
```

Comment:
**nit:** I felt this block is hard to understand as why `rml` copied to `results` first and again became empty. Can't we just do something like

```
if opts.name == "" {
  return rml
}
results := RemoteMixinList{}
for _, rmi := range rml {
  if strings.Contains(rmi.Name, opts.Name) {
    results = append(results, rmi)
  }
}
sort.Sort(results)
return results
```

Reply:
Much simpler and easier to understand, agreed.  Thank you!  Updated.